README
Stjerm is a GTK+-based drop-down terminal emulator similar to the consoles used
PC games such as Quake and Half-Life 2.  Stjerm sets itself apart from
similar programs by providing a minimalistic interface combined with a small 
file size, lightweight memory usage and easy integration with composite window
managers, such as Compiz.

Information about the current version of Stjerm (including new features and 
changes) can be found in the file NEWS.

For building and installation instructions, please see the file INSTALL.

Dependencies for building:  
  * pkg-config
  * GTK+ 2 (>= 2.10)
  * VTE (>= 0.16)

The two subdirectories, 'src' and 'doc', contain the source code files and 
manual for Stjerm, respectively.

Stjerm is distributed under the terms of the GNU General Public License, 
version 2.  For complete information, see the file COPYING.


NEWS
NEWS file for Stjerm <http://code.google.com/p/stjerm-terminal-emulator/>

0.14 (2011-01-30)
  * New feature: You can now set whether you want an audible bell enabled for
    the terminal though the .Xdefaults and runtime arguments bell and -bell, 
    respectively.  The default value is TRUE.  See the man page for more info.
  * Fixed a bug where the command line argument for setting the amount of
    scrollback lines was set to '-bl' instead of '-l'. 
  * Changed '--v' to '-v' for version info since a '--' is used for a long 
    version of a command, not a short version.  '--version' is still there and
    remains unchaged.
  * Stjerm's source code is now released as both a bzip2 and xz compressed 
    tarball.  Both contain the same source code, however the xz tarball is
    slightly smaller in size.
  * The man page and --help runtime argument have had slight updates to them.

0.13 (2009-06-24):
  * Man page has been rewritten making it cleaner and more concise
  * --help and --info options have been alphabetized making it easier to read
  * Removed version.m4 as it was superfluous (The version number is updated in 
    cofigure.ac)
  * An RPM is now available for Fedora 11 users!

0.12 (2009-03-06):
  * Added copy and paste keyboard shortcuts 'c' and 'v', respectively
  * Keyboard shortcuts to switch to the previous and next tab have been 
    re-mapped to page up and page down, respectively
  * Added version information output through --version and --v
  * Tabs now loop when the end or beginning of the tab bar is reached
  * Many files and functions have been updated to conform to GNU standards (Such 
    as README and print_version)
  * Added NEWS and TODO
  * Version information is now updated through version.m4


src/main.c
/*
 * main.c
 * This file is part of Stjerm
 *
 * Copyright (C) 2007 - Stjepan Glavina
 * Copyright (C) 2007 - Markus Groß
 * Copyright (C) 2008, 2009, 2011 Thomas Pifer
 *
 * Stjerm is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Stjerm is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Stjerm; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */

#include <gtk/gtk.h>
#include "stjerm.h"

int sargc;
char **sargv;

extern Option options[OPTION_COUNT]; //Defined in config.c

void print_help(void);
void print_info(void);
void print_version(void);

void print_help(void) {
    printf("Usage: %s [action] [options]\n\n"
           "Actions are:\n"
           "  --help\tPrint this message\n"
           "  --info\tPrint .Xdefaults options and exit\n"
           "  --version\tPrint version information and exit\n"
           "  --toggle\tShow/Hide the current stjerm instance\n"
           "(When using an action command, the options are ignored)\n\n"
           "Options are:\n", sargv[0]);
    /*
     * i is declared outside of the for statement because the current C standard 
     * that GCC uses does not support declaration inside of a loop.
     */
    int i = 0;
    Option o;   //Option is a struct defined in stjerm.h
    for (i = 0; i < OPTION_COUNT; i++) {
        o = options[i];
        printf("  %s %s\t%s\n", o.short_name, o.var_type, o.desc);
    }
    printf("\t\tYou may specify no palette, or a complete one with 16 total colors\n"
           "\t\tFor this you have to use -c0, -c1, ..., -c15.\n\n"
           "Note:     Any colors in hex codes are entered without the leading '#'.\n\n"
           "Examples: %s -fn \"Liberation Mono 11\" -bg white -k F12 -sh /bin/bash\n"
           "          %s -bg black -fg green -w 800 -h 405 -s right -m alt -k f -l 10000\n"
           "          %s -b thick -w 500 -p left -t best -o 90 -m shift -k tab -sh /bin/zsh\n\n"
           "Use --info to get a list of available .Xdefaults options for stjerm.\n\n"
           "Report bugs to zero456@gmail.com.\n"
           "stjerm home page: http://code.google.com/p/stjerm-terminal-emulator/\n"
           "General help using GNU software: http://www.gnu.org/gethelp/\n",
           sargv[0], sargv[0], sargv[0]);
}

void print_info(void) {
    printf("Usage: %s [action] [options]\n\n"
           "Actions are:\n"
           "  --help\tDisplay command line options menu\n"
           "  --info\tDisplay .Xdefault options\n"
           "  --toggle\tShow/Hide the current stjerm instance\n"
           "(When using an action command, the .Xdefaults options are ignored)\n\n"
           "The .Xdefault options are read from ~/.Xdefaults.\n"
           "To specify an option in .Xdefaults use this syntax:\n"
           "stjerm.OPTION: VALUE\n"
           "Example: stjerm.key: f12\n\n"
           "Available .Xdefaults options:\n", sargv[0]);
    int i = 0;
    Option o;
    for (i = 0; i < OPTION_COUNT; i++) {
        o = options[i];
        printf("  %s:  \t%s      \t%s\n", o.long_name, o.var_type, o.desc);
    }
    printf("\t\t\t\tYou may specify no palette, or a complete one with 16 total colors.\n"
           "\t\t\t\tFor this you have to use color0, color1, ..., color15.\n");
}

void print_version(void) {
    printf("Stjerm %s\n"
           "Copyright (C) 2007 Stjepan Glavina\n"
           "Copyright (C) 2007 Markus Gross\n"
           "Copyright (C) 2011 Thomas Pifer\n"
           "License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>\n"
           "This is free software: you are free to change and redistribute it.\n"
           "There is NO WARRANTY, to the extent permitted by law.\n", 
           STJERM_VERSION);
}

int main(int argc, char *argv[]) {
    sargc = argc;
    sargv = argv;

    if (!g_thread_supported())
        g_thread_init(NULL);
    gdk_threads_init();
    gtk_init(&argc, &argv);

    conf_init();
    build_mainwindow();

    gdk_threads_enter();
    gtk_main();
    gdk_threads_leave();

    return 0;
}



src/stjerm.h
/*
 * stjerm.h
 * This file is part of Stjerm
 *
 * Copyright (C) 2007 - Stjepan Glavina
 * Copyright (C) 2007 - Markus Groß
 * Copyright (C) 2008, 2009, 2011 Thomas Pifer
 *
 * Stjerm is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Stjerm is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Stjerm; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */

#ifndef __STJERM_H__
#define __STJERM_H__

#include <gtk/gtk.h>
#include <X11/Xlib.h>

#define TERM_ICONIFY_WINDOW    0
#define TERM_DEICONIFY_WINDOW  1
#define TERM_RAISE_WINDOW      2
#define TERM_LOWER_WINDOW      3
#define TERM_MAXIMIZE_WINDOW   4
#define TERM_RESTORE_WINDOW    5
#define TERM_REFRESH_WINDOW    6
#define TERM_RESIZE_WINDOW     7
#define TERM_MOVE_WINDOW       8

#define BORDER_THIN  0
#define BORDER_THICK 1
#define BORDER_NONE  2

#define POS_TOP         0
#define POS_BOTTOM      1
#define POS_LEFT        2
#define POS_RIGHT       3
#define POS_TOPLEFT     4
#define POS_TOPRIGHT    5
#define POS_BOTTOMLEFT  6
#define POS_BOTTOMRIGHT 7

#define TABS_NEVER  0
#define TABS_ONE    1
#define TABS_ALWAYS 2

typedef struct {
    char long_name[12];
    char short_name[12];
    char var_type[12];
    char desc[150];
} Option;

#define OPTION_COUNT 26

extern void print_help(void);
extern void print_info(void);
extern void print_version(void);

extern void build_mainwindow(void);
extern GtkWidget* build_term(void);
extern void build_popupmenu(void);

extern void mainwindow_toggle(int sig);
extern void mainwindow_create_tab(void);
extern void mainwindow_close_tab(void);
extern void mainwindow_toggle_full(void);
extern void init_key(void);
extern void grab_key(void);
extern void wait_key(void);

extern void conf_init(void);
extern char* conf_get_font(void);
extern float conf_get_opacity(void);
extern GdkColor conf_get_bg(void);
extern GdkColor conf_get_fg(void);
extern int conf_get_border(void);
extern unsigned int conf_get_mod(void);
extern KeySym conf_get_key(void);
extern int conf_get_width(void);
extern int conf_get_height(void);
extern void conf_get_position(int*, int*);
extern int conf_get_scrollbar(void);
extern char* conf_get_shell(void);
extern int conf_get_lines(void);
extern int conf_get_show_tab(void);
extern char* conf_get_term_name(void);
extern GtkPositionType conf_get_tab_pos(void);
extern GdkColor* conf_get_color_palette(void);
extern gboolean conf_get_tab_fill(void);
extern gboolean conf_get_allow_bold(void);
extern GdkModifierType conf_get_key_mod(void);
extern gboolean conf_get_auto_hide(void);
extern char* conf_get_bg_image(void);
extern gboolean conf_get_scroll_on_output();
extern gboolean conf_get_bell();
extern int conf_get_cursor_blink(void);
extern int conf_get_cursor_shape(void);

#endif /* __STJERM_H__ */



src/config.c
/*
 * config.c
 * This file is part of Stjerm
 *
 * Copyright (C) 2007 - Stjepan Glavina
 * Copyright (C) 2007 - Markus Groß
 * Copyright (C) 2008, 2009, 2011 Thomas Pifer
 * 
 * Stjerm is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Stjerm is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Stjerm; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */

#include <gtk/gtk.h>
#include <gdk/gdkx.h>
#include <X11/Xlib.h>
#include <X11/Xresource.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/stat.h>
#include <unistd.h>
#include <pwd.h>
#include <signal.h>
#include <vte/vte.h>
#include "stjerm.h"

extern int sargc;
extern char **sargv;
extern GtkWidget *mainwindow;
static char _font[200];
static float _opacity;
static GdkColor _bg;
static GdkColor _fg;
static int _border;
static unsigned int _mod;
static KeySym _key;
static int _width;
static int _height;
static int _pos;
static int _posx;
static int _posy;
static int _scrollpos;
static char _shell[200];
static int _lines;
static int _showtab;
static char _termname[200];
static GtkPositionType _tabpos;
static GdkColor _palette[16];
static int read_colors;
static gboolean _tabfill;
static gboolean _allowbold;
static GdkModifierType _keymod;
static gboolean _autohide;
static char _bgimage[200];
static gboolean _scrolloutput;
static gboolean _bell;
static int _cursorblink;
static int _cursorshape;

static void set_border(char*);
static void set_mod(char*);
static void set_key(char*);
static void set_pos(char *v);
static void set_cursor_blink(char*);
static void set_cursor_shape(char*);
static GtkPositionType read_pos(char *v);
static gboolean parse_hex_color(char *value, GdkColor *color);
static gboolean parse_bool_str(char *value, gboolean def);
static GdkModifierType parse_mod(char *value);
static pid_t get_stjerm_pid(void);

void read_value(char *name, char *value);
void init_default_values(void);
void conf_init(void);
char* conf_get_font(void);
float conf_get_opacity(void);
GdkColor conf_get_bg(void);
GdkColor conf_get_fg(void);
int conf_get_border(void);
unsigned int conf_get_mod(void);
KeySym conf_get_key(void);
int conf_get_width(void);
int conf_get_height(void);
void conf_get_position(int*, int*);
int conf_get_scrollbar(void);
char* conf_get_shell(void);
int conf_get_lines(void);
int conf_get_show_tab(void);
char* conf_get_term_name(void);
GtkPositionType conf_get_tab_pos(void);
GdkColor* conf_get_color_palette(void);
gboolean conf_get_tab_fill(void);
gboolean conf_get_allow_bold(void);
GdkModifierType conf_get_key_mod(void);
gboolean conf_get_auto_hide(void);
char* conf_get_bg_image(void);
gboolean conf_get_scroll_on_output(void);
int conf_get_cursor_blink(void);
int conf_get_cursor_shape(void);

Option options[OPTION_COUNT] = {
        { "allowbold", "-ab", "BOOLEAN", "Allow bold fonts or not. Default: true." },
        { "autohide", "-ah", "BOOLEAN", "Whether or not to hide stjerm when it looses focus. Default: true." },
        { "background", "-bg", "COLOR", "Background color. Default: Black." },
        { "bell", "-bell", "BOOLEAN", "Whether or not the terminal will make an audible bell. Default: True."},
        { "bgimage", "-bgimg", "FILE", "Background image to use on terminal." },
        { "border", "-b", "TYPE", "Border type: thin, thick, none. Default: none." },
        { "colorX", "-cX", "COLOR", "Specify color X of the terminals color palette" },
        { "font", "-fn", "FONT", "Terminal font and size (eg: Sans 10). Default: Bistream Vera Sans 10." },
        { "foreground", "-fg", "COLOR", "Foreground color. Default: White." },
        { "height", "-h", "NUMBER", "Window height. Default: 400." },
        { "key", "-k", "KEY", "Shortcut key (eg: f12)." },
        { "keymod", "-km", "MODIFIER", "Modifier for keyboard shortcuts. Can be a combination (with +) of modifiers (eg: shift+control)." },
        { "lines", "-l", "NUMBER", "Scrollback lines. 0 to disable scrollback. Default: 1000." },
        { "mod", "-m", "MODIFIER", "Meta modifier key: shift, control, alt, windows, none." },
        { "opacity", "-o", "NUMBER", "Opacity (range: 10 - 100). Default: 100." },
        { "position", "-p", "POSITION", "Window position: top, bottom, left, right. Default: top." },
        { "scrollbar", "-s", "POSITION", "Scrollbar position: left, right, none. Default: none." },
        { "scroll", "-sc", "BOOLEAN", "Whether to scroll the terminal on output. Default: true." },
        { "shell", "-sh", "STRING", "Terminal Shell. Default: the user's default shell." },
        { "showtab", "-st", "VALUE", "Tabbar visibility (one: only visible when > 1 tabs): never, one, always." },
        { "tabfill", "-tf", "BOOLEAN", "Whether tabs fill whole tabbar space. Default: true." },
        { "tablabel", "-tl", "STRING", "Label of the tabs. Default: term." },
        { "tabpos", "-tp", "POSITION", "Tabbar position: top, bottom, left, right. Default: bottom." },
        { "width", "-w", "NUMBER", "Window width. Default: 800." },
		{ "cursorblink", "-cb", "STRING", "Cursor blink: system, on, off. Default: system."},
		{ "cursorshape", "-cs", "STRING", "Cursor shape: block, ibeam, underline. Default: block."}
};

pid_t get_stjerm_pid(void) {
    FILE *p = popen("pidof stjerm", "r");
    if (p == NULL) {
        fprintf(stderr, "ERROR: Unable to get stjerm pid\n");
        exit(1);
    }
    char buffer[100];
    fgets(buffer, sizeof(buffer), p);
    pclose(p);
    char **list;
    list = g_strsplit_set(buffer, " ", -1);
    int i = 0;
    while (list[i] != NULL)
        i++;
    pid_t pid = (pid_t) atoi(list[i - 1]);
    g_strfreev(list);
    return pid;
}

void set_border(char *v) {
    if (!strcmp(v, "thin"))
        _border = BORDER_THIN;
    else if (!strcmp(v, "thick"))
        _border = BORDER_THICK;
    else
        _border = BORDER_NONE;
}

void set_mod(char *v) {
    v[0] = tolower(v[0]);
    if (!strcmp(v, "shift"))
        _mod = ShiftMask;
    else if (!strcmp(v, "control"))
        _mod = ControlMask;
    else if (!strcmp(v, "ctrl"))
        _mod = ControlMask;
    else if (!strcmp(v, "alt"))
        _mod = Mod1Mask;
    else if (!strcmp(v, "mod1"))
        _mod = Mod1Mask;
    else if (!strcmp(v, "windows"))
        _mod = Mod4Mask;
    else if (!strcmp(v, "win"))
        _mod = Mod4Mask;
    else
        _mod = 0;
}
/*
 * Set key character that will show/hide Stjerm
 */
void set_key(char *v) {
    _key = 0;

    v[0] = tolower(v[0]);
    _key = XStringToKeysym(v);

    if (!_key) {
        v[0] = toupper(v[0]);
        _key = XStringToKeysym(v);
    }
}

void set_pos(char *v) {
    if (!strcmp(v, "top"))
        _pos = POS_TOP;
    else if (!strcmp(v, "bottom"))
        _pos = POS_BOTTOM;
    else if (!strcmp(v, "left"))
        _pos = POS_LEFT;
    else if (!strcmp(v, "right"))
        _pos = POS_RIGHT;
    else if (!strcmp(v, "topleft"))
        _pos = POS_TOPLEFT;
    else if (!strcmp(v, "topright"))
        _pos = POS_TOPRIGHT;
    else if (!strcmp(v, "bottomleft"))
        _pos = POS_BOTTOMLEFT;
    else if (!strcmp(v, "bottomright"))
        _pos = POS_BOTTOMRIGHT;
}

GtkPositionType read_pos(char *v) {
    if (!strcmp(v, "top"))
        return GTK_POS_TOP;
    else if (!strcmp(v, "left"))
        return GTK_POS_LEFT;
    else if (!strcmp(v, "right"))
        return GTK_POS_RIGHT;
    else
        return GTK_POS_BOTTOM;
}

void set_cursor_blink(char *v) {
    if (!strcmp(v, "system"))
        _cursorblink = VTE_CURSOR_BLINK_SYSTEM;
    else if (!strcmp(v, "on"))
        _cursorblink = VTE_CURSOR_BLINK_ON;
    else if (!strcmp(v, "off"))
        _cursorblink = VTE_CURSOR_BLINK_OFF;
}

void set_cursor_shape(char *v) {
    if (!strcmp(v, "block"))
        _cursorshape = VTE_CURSOR_SHAPE_BLOCK;
    else if (!strcmp(v, "ibeam"))
        _cursorshape = VTE_CURSOR_SHAPE_IBEAM;
    else if (!strcmp(v, "underline"))
        _cursorshape = VTE_CURSOR_SHAPE_UNDERLINE;
}

gboolean parse_hex_color(char *value, GdkColor *color) {
    if (!gdk_color_parse(value, color)) {
        char *value2 = g_strconcat("#", value, NULL);
        gboolean res = gdk_color_parse(value2, color);
        free(value2);
        return res;
    } else
        return TRUE;
}

gboolean parse_bool_str(char *value, gboolean def) {
    g_strstrip(value);
    gboolean res = def;
    if (!strcasecmp("false", value) || !strcasecmp("0", value) 
            || !strcasecmp("no", value))
        res = FALSE;
    if (!strcasecmp("true", value) || !strcasecmp("1", value) 
            || !strcasecmp("yes", value))
        res = TRUE;
    return res;
}

GdkModifierType parse_mod(char *value) {
    g_strstrip(value);
    if (!strcasecmp("control", value) || !strcasecmp("ctrl", value))
        return GDK_CONTROL_MASK;
    else if (!strcasecmp("alt", value))
        return GDK_MOD1_MASK;
    else if (!strcasecmp("shift", value))
        return GDK_SHIFT_MASK;
    else if (!strcasecmp("windows", value))
        return GDK_SUPER_MASK;
    else
        return 0;
}

void init_default_values(void) {
    strcpy(_font, "Bitstream Vera Sans Mono 10");
    gdk_color_parse("black", &_bg);
    gdk_color_parse("white", &_fg);
    _scrollpos = -1;
    _border = BORDER_NONE;
    _opacity = 100.0f;
    strcpy(_bgimage, "");
    _width = 800;
    _height = 400;
    _pos = POS_TOP;
    _mod = 0;
    _key = 0;
    strcpy(_shell, getpwuid(getuid())->pw_shell);
    _lines = 1000;
    _showtab = TABS_ONE;
    _tabpos = GTK_POS_BOTTOM;
    strcpy(_termname, "term");
    read_colors = 0;
    _tabfill = FALSE;
    _allowbold = TRUE;
    _keymod = GDK_CONTROL_MASK | GDK_SHIFT_MASK;
    _autohide = TRUE;
    _scrolloutput = TRUE;
    _bell = TRUE;
	_cursorblink = VTE_CURSOR_BLINK_SYSTEM;
	_cursorshape = VTE_CURSOR_SHAPE_BLOCK;
}

void read_value(char *name, char *value) {

//The *name and *value are derived from sargv which is given one by one through 
//a loop from conf_init()

    if (name != NULL && value != NULL) {
        if (name[0] == '#')
            return;
        g_strstrip(name);
        g_strstrip(value);
        if (!strcmp("font", name) || !strcmp("-fn", name))
            strcpy(_font, value);
        else if (!strcmp("background", name) || !strcmp("-bg", name)) {
            if (!parse_hex_color(value, &_bg))
                gdk_color_parse("black", &_bg);
        } else if (!strcmp("foreground", name) || !strcmp("-fg", name)) {
            if (!parse_hex_color(value, &_fg))
                gdk_color_parse("white", &_fg);
        } else if (!strcmp("scrollbar", name) || !strcmp("-s", name)) {
            if (!strcmp(value, "true"))
                _scrollpos = POS_RIGHT;
            else if (!strcmp(value, "left"))
                _scrollpos = POS_LEFT;
            else if (!strcmp(value, "right"))
                _scrollpos = POS_RIGHT;
            else
                _scrollpos = -1;
        } else if (!strcmp("border", name) || !strcmp("-b", name))
            set_border(value);
        else if (!strcmp("opacity", name) || !strcmp("-o", name))
            _opacity = atof(value);
        else if (!strcmp("bgimage", name) || !strcmp("-bgimg", name))
            strcpy(_bgimage, value);
        else if (!strcmp("width", name) || !strcmp("-w", name))
            _width = atoi(value);
        else if (!strcmp("height", name) || !strcmp("-h", name))
            _height = atoi(value);
        else if (!strcmp("position", name) || !strcmp("-p", name))
            set_pos(value);
        else if (!strcmp("mod", name) || !strcmp("-m", name))
            set_mod(value);
        else if (!strcmp("key", name) || !strcmp("-k", name))
            set_key(value);
        else if (!strcmp("shell", name) || !strcmp("-sh", name))
            strcpy(_shell, value);
        else if (!strcmp("lines", name) || !strcmp("-l", name))
            _lines = atoi(value);
        else if (!strcmp("showtab", name) || !strcmp("-showtab", name)) {
            if (!strcasecmp(value, "always"))
                _showtab = TABS_ALWAYS;
            else if (!strcasecmp(value, "never"))
                _showtab = TABS_NEVER;
        } else if (!strcmp("tabpos", name) || !strcmp("-tabpos", name))
            _tabpos = read_pos(value);
        else if (!strcmp("tablabel", name) || !strcmp("-tablabel", name))
            strcpy(_termname, value);
		else if (!strcmp("cursorblink", name) || !strcmp("-cb", name))
            set_cursor_blink(value);
		else if (!strcmp("cursorshape", name) || !strcmp("-cs", name))
            set_cursor_shape(value);
        else if (g_str_has_prefix(name, "color") || g_str_has_prefix(name, "-c")) {
            g_strcanon(name, "0123456789", ' ');
            g_strchug(name);
            parse_hex_color(value, &_palette[atoi(name)]);
            read_colors++;
        } else if (!strcmp("tabfill", name) || !strcmp("-tf", name))
            _tabfill = parse_bool_str(value, _tabfill);
        else if (!strcmp("allowbold", name) || !strcmp("-ab", name))
            _allowbold = parse_bool_str(value, _allowbold);
        else if (!strcmp("keymod", name) || !strcmp("-km", name)) {
            char **list;
            list = g_strsplit_set(value, "+", -1);
            GdkModifierType tmp = 0;
            int i = 0;
            while (list[i] != NULL)
                tmp = tmp | parse_mod(list[i++]);
            g_strfreev(list);
            if (tmp != 0)
                _keymod = tmp;
        } else if (!strcmp("autohide", name) || !strcmp("-ah", name))
            _autohide = parse_bool_str(value, _autohide);
        else if (!strcmp("scroll", name) || !strcmp("-sc", name))
            _scrolloutput = parse_bool_str(value, _scrolloutput);
        else if (!strcmp("bell", name) || !strcmp("-bell", name))
			if (!strcasecmp(value, "false"))
				_bell = FALSE;
    }
}

void conf_init(void) {
    init_default_values();
    gboolean keyoption = FALSE;

    Display *dpy = gdk_x11_display_get_xdisplay(gdk_display_get_default());
    XrmInitialize();
    char basename[12] = "/.Xdefaults";
    char *filename = strcat(getpwuid(getuid())->pw_dir, basename);
    XrmSetDatabase(dpy, XrmGetFileDatabase(filename));

    char *op;
    int i;
    Option o;
    for (i = 0; i < OPTION_COUNT; i++) {
        o = options[i];
        if ((op = XGetDefault(dpy, "stjerm", o.long_name)))
            read_value(o.long_name, op);
    }
    char color[8];
    for (i = 0; i < 16; i++) {
        sprintf(color, "color%d", i);
        if ((op = XGetDefault(dpy, "stjerm", color)))
            read_value(color, op);
    }

    for (i = 1; i < sargc; i++) {
        if (sargv != NULL) {
            if (!strcmp(sargv[i], "--help")) {
                print_help();
                exit(1);
            } else if (!strcmp(sargv[i], "--info")) {
                print_info();
                exit(1);
            } else if (!strcmp("--toggle", sargv[i])) {
                kill(get_stjerm_pid(), SIGUSR1);
                exit(1);
            } else if (!strcmp("--version", sargv[i]) || !strcmp("-v", sargv[i])){
	     	    print_version();
		        exit(1);
	        }
        }
        if (i + 1>= sargc)
            break;

        read_value(sargv[i], sargv[i + 1]);
    }

    if (keyoption == FALSE && _key == 0) {
        printf("Hint: You started stjerm without specifying a shortcut key\n"
               "      To show/hide stjerm run stjerm with the toggle option\n"
               "      like this: stjerm --toggle");
    }

    struct stat st;
    if (stat(_shell, &st) != 0) {
        fprintf(stderr, "ERROR: The shell '%s' can't be opened\n", _shell);
        exit(1);
    }

    if (_lines < 0) {
        fprintf(stderr, "ERROR: A scrollback line count < 0 is not allowed\n");
        exit(1);
    }

    if (keyoption == TRUE && _key == 0) {
        fprintf(stderr, "ERROR: Wrong shortcut key is defined\n");
        exit(1);
    }

    if (read_colors > 0 && read_colors < 16) {
        fprintf(stderr, "ERROR: Read only %d colors, not 16\n"
                "Specify a complete color palette\n", read_colors);
        exit(1);
    }

    signal(SIGUSR1, (__sighandler_t) mainwindow_toggle);

    int scrw = gdk_screen_get_width(gdk_screen_get_default());
    int scrh = gdk_screen_get_height(gdk_screen_get_default());

    if (_pos == POS_TOP) {
        _posx = (scrw - _width) / 2;
        _posy = 0;
    } else if (_pos == POS_BOTTOM) {
        _posx = (scrw - _width) / 2;
        _posy = scrh - _height;
    } else if (_pos == POS_LEFT) {
        _posx = 0;
        _posy = (scrh - _height) / 2;
    } else if (_pos == POS_RIGHT) {
        _posx = scrw - _width;
        _posy = (scrh - _height) / 2;
    } else if (_pos == POS_TOPLEFT) {
        _posx = 0;
        _posy = 0;
    } else if (_pos == POS_TOPRIGHT) {
        _posx = scrw - _width;
        _posy = 0;
    } else if (_pos == POS_BOTTOMLEFT) {
        _posx = 0;
        _posy = scrh - _height;
    } else if (_pos == POS_BOTTOMRIGHT) {
        _posx = scrw - _width;
        _posy = scrh - _height;
    }
}

char* conf_get_font(void) {
    return _font;
}

float conf_get_opacity(void) {
    return _opacity;
}

GdkColor conf_get_bg(void) {
    return _bg;
}

GdkColor conf_get_fg(void) {
    return _fg;
}

int conf_get_border(void) {
    return _border;
}

unsigned int conf_get_mod(void) {
    return _mod;
}

KeySym conf_get_key(void) {
    return _key;
}

int conf_get_width(void) {
    return _width;
}

int conf_get_height(void) {
    return _height;
}

void conf_get_position(int *x, int *y) {
/*
 * '_posx' and '_posy' are set when 'conf_init is run'
 */
    *x = _posx;
    *y = _posy; 
}

int conf_get_scrollbar(void) {
    return _scrollpos;
}

char* conf_get_shell(void) {
    return _shell;
}

int conf_get_lines(void) {
    return _lines;
}

int conf_get_show_tab(void) {
    return _showtab;
}

char* conf_get_term_name(void) {
    return _termname;
}

GtkPositionType conf_get_tab_pos(void) {
    return _tabpos;
}

GdkColor* conf_get_color_palette(void) {
    if (read_colors <= 0)
        return NULL;
    else
        return _palette;
}

gboolean conf_get_tab_fill(void) {
    return _tabfill;
}

gboolean conf_get_allow_bold(void) {
    return _allowbold;
}

GdkModifierType conf_get_key_mod(void) {
    return _keymod;
}

gboolean conf_get_auto_hide(void) {
    return _autohide;
}

char* conf_get_bg_image(void) {
    if (!strcmp(_bgimage, ""))
        return NULL;
    else
        return _bgimage;
}

gboolean conf_get_scroll_on_output(void) {
    return _scrolloutput;
}

gboolean conf_get_bell(void) {
    return _bell;
}

int conf_get_cursor_blink(void) {
    return _cursorblink;
}

int conf_get_cursor_shape(void) {
    return _cursorshape;
}


